modules/core/src/main/java/org/apache/ignite/ssl/AbstractSslContextFactory.java

```diff
@@ -0,0 +1,227 @@
+    }
+
+    /**
+     * Gets enabled cipher suites.
+     *
+     * @return enabled cipher suites
+     */
+    public String[] getCipherSuites() {
+        return cipherSuites;
+    }
+
+    /**
+     * Gets enabled protocols.
+     *
+     * @return Enabled protocols.
+     */
+    public String[] getProtocols() {
+        return protocols;
+    }
+
+    /**
+     * Sets enabled protocols.
+     *
+     * @param protocols Enabled protocols.
+     */
+    public void setProtocols(String... protocols) {
+        this.protocols = protocols;
+    }
+
+    /**
+     * Creates SSL context based on factory settings.
+     *
+     * @return Initialized SSL context.
+     * @throws SSLException If SSL context could not be created.
+     */
+    private SSLContext createSslContext() throws SSLException {
+        checkParameters();
+
+        KeyManager[] keyMgrs = createKeyManagers();
+
+        TrustManager[] trustMgrs = createTrustManagers();
+
+        try {
+            SSLContext ctx = SSLContext.getInstance(proto);
+
+            if (cipherSuites != null || protocols != null) {
+                SSLParameters sslParameters = new SSLParameters();
+
+                if (cipherSuites != null)
+                    sslParameters.setCipherSuites(cipherSuites);
+
+                if (protocols != null)
+                    sslParameters.setProtocols(protocols);
+
+                ctx = new SSLContextWrapper(ctx, sslParameters);
+            }
+
+            ctx.init(keyMgrs, trustMgrs, null);
+
+            return ctx;
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new SSLException("Unsupported SSL protocol: " + proto, e);
+        }
+        catch (KeyManagementException e) {
+            throw new SSLException("Failed to initialized SSL context.", e);
+        }
+    }
+
+    /**
+     * @param param Value.
+     * @param name Name.
+     * @throws SSLException If {@code null}.
+     */
+    protected void checkNullParameter(Object param, String name) throws SSLException {
+        if (param == null)
+            throw new SSLException("Failed to initialize SSL context (parameter cannot be null): " + name);
+    }
+
+    /**
+     * Checks that all required parameters are set.
+     *
+     * @throws SSLException If any of required parameters is missing.
+     */
+    protected abstract void checkParameters() throws SSLException;
+
+    /**
+     * @return Created Key Managers.
+     * @throws SSLException If Key Managers could not be created.
+     */
+    protected abstract KeyManager[] createKeyManagers() throws SSLException;
+
+    /**
+     * @return Created Trust Managers.
+     * @throws SSLException If Trust Managers could not be created.
+     */
+    protected abstract TrustManager[] createTrustManagers() throws SSLException;
+
+    /**
+     * Disabled trust manager, will skip all certificate checks.
+     */
+    protected static class DisabledX509TrustManager implements X509TrustManager {
```

Comment:
Do we really need it in abstract class? It used only by SslContextFactory static method.